### ISSUE TITLE
Question about preferred way to iterate StableGraph

### ISSUE THREAD

**Issue by jgarvin** [petgraph/petgraph#387]
This isn't a bug or feature request but just a question, please tell me if there is a better forum. Here is a code snippet from a project of mine, that is iterating over a `StableGraph`, and running a user provided callback on each edge weight:

```
    for e in graph.edges_directed(source, dir) {
        let target_id = e.id();
        let target = match dir {
            petgraph::Direction::Incoming => e.source(),
            petgraph::Direction::Outgoing => e.target(),
        };
        let target = *index_to_id
            .get(&target)
            .ok_or(NodeWithoutEntityError(target.index()))?;
        f(graph.edge_weight(target_id).unwrap(), target)?;
    }
```

This works fine. Now I have an almost identical function that wants to let users mutate the edge weights:

```
    for e in graph.edges_directed(source, dir) {
        let target_id = e.id();
        let target = match dir {
            petgraph::Direction::Incoming => e.source(),
            petgraph::Direction::Outgoing => e.target(),
        };
        let target = *index_to_id
            .get(&target)
            .ok_or(NodeWithoutEntityError(target.index()))?;
        f(graph.edge_weight_mut(target_id).unwrap(), target)?;
    }
```

This one doesn't compile because `edges_directed` immutably borrows the graph, and `edge_weight_mut` wants to do a mutable borrow.

The easiest work around is for me to iterate once and collect everything iterated into a separate container, then iterate that container so that that container is borrowed instead and doesn't interfere with me calling `edge_weight_mut`. However this is very inefficient, both because I'm iterating twice and because I'm allocating a Vec. Alternatively I could make the edge weights RefCells, but this also carries an overhead.

Is there a better way? I am only trying to mutate the fields inside my edge weight type, I am not trying to change the actual structure of the graph itself while iterating. If there were a `edges_directed_mut` would it solve this problem?

**Comment by bluss** [comment]
There's a non-borrowing way to do this traversal, using https://docs.rs/petgraph/0.5.1/petgraph/stable_graph/struct.StableGraph.html#method.neighbors_directed  then .detach() and the edge stepping method on that Walker. I think it can be used, question is if it is convenient enough.

**Comment by jgarvin** [comment]
Ah, that looks perfect. Looking at the implementation it looks like it uses a reference to a slice rather than a separate container, so I assume this doesn't allocate? i.e. it's doing something more clever than just wrapping up the Vec workaround? Might be helpful for docs to mention if detach is O(1).

**Comment by jgarvin** [comment]
Thanks that worked, here's the code for anyone else who has this question in the future:

    let mut neighbors = graph.neighbors_directed(source, dir).detach();

    while let Some((edge_id, target_id)) = neighbors.next(graph) {
        let target = *index_to_id
            .get(&target_id)
            .ok_or(NodeWithoutEntityError(target_id.index()))?;
        f(&mut graph[edge_id], target)?;
    }
